File: frontend/src/components/benchmark/content/floyd_warshall/implementations/floyd_warshall.js
class FloydWarshall {
    constructor() {
        this.nodes = new Map();
        this.dist = new Map();
        this.nextNode = new Map();
    }

    addEdge(u, v, weight) {
        if (!this.nodes.has(u)) this.nodes.set(u, new Set());
        if (!this.nodes.has(v)) this.nodes.set(v, new Set());
        this.nodes.get(u).add(v);
        this.nodes.get(v).add(u);

        if (!this.dist.has(u)) this.dist.set(u, new Map());
        if (!this.dist.has(v)) this.dist.set(v, new Map());
        this.dist.get(u).set(v, weight);
        this.dist.get(v).set(u, weight);

        if (!this.nextNode.has(u)) this.nextNode.set(u, new Map());
        if (!this.nextNode.has(v)) this.nextNode.set(v, new Map());
        this.nextNode.get(u).set(v, v);
        this.nextNode.get(v).set(u, u);
    }

    addDirectedEdge(u, v, weight) {
        if (!this.nodes.has(u)) this.nodes.set(u, new Set());
        if (!this.nodes.has(v)) this.nodes.set(v, new Set());
        this.nodes.get(u).add(v);

        if (!this.dist.has(u)) this.dist.set(u, new Map());
        if (!this.dist.has(v)) this.dist.set(v, new Map());
        this.dist.get(u).set(v, weight);

        if (!this.nextNode.has(u)) this.nextNode.set(u, new Map());
        if (!this.nextNode.has(v)) this.nextNode.set(v, new Map());
        this.nextNode.get(u).set(v, v);
    }

    buildFromEdges(edges, directed = false) {
        for (const [u, v, weight] of edges) {
            if (directed) {
                this.addDirectedEdge(u, v, weight);
            } else {
                this.addEdge(u, v, weight);
            }
        }
    }

    getAllNodes() {
        return Array.from(this.nodes.keys());
    }

    initialize() {
        const allNodes = this.getAllNodes();

        for (const u of allNodes) {
            if (!this.dist.has(u)) this.dist.set(u, new Map());
            if (!this.nextNode.has(u)) this.nextNode.set(u, new Map());

            for (const v of allNodes) {
                if (u === v) {
                    this.dist.get(u).set(v, 0);
                } else if (!this.dist.get(u).has(v)) {
                    this.dist.get(u).set(v, Infinity);
                }
            }
        }
    }

    allPairsShortestPaths() {
        this.initialize();
        const allNodes = this.getAllNodes();

        for (const k of allNodes) {
            for (const i of allNodes) {
                for (const j of allNodes) {
                    if (this.dist.get(i).get(k) !== Infinity &&
                        this.dist.get(k).get(j) !== Infinity &&
                        this.dist.get(i).get(k) + this.dist.get(k).get(j) < this.dist.get(i).get(j)) {
                        this.dist.get(i).set(j, this.dist.get(i).get(k) + this.dist.get(k).get(j));
                        this.nextNode.get(i).set(j, this.nextNode.get(i).get(k));
                    }
                }
            }
        }
    }

    shortestPath(u, v) {
        if (!this.dist.has(u) || !this.dist.get(u).has(v)) {
            return Infinity;
        }
        return this.dist.get(u).get(v);
    }

    getPath(u, v) {
        if (!this.nextNode.has(u) || !this.nextNode.get(u).has(v)) {
            return [];
        }

        const path = [u];

        while (u !== v) {
            u = this.nextNode.get(u).get(v);
            path.push(u);
        }

        return path;
    }

    getAllDistances() {
        return this.dist;
    }

    hasNegativeCycle() {
        const allNodes = this.getAllNodes();

        for (const i of allNodes) {
            if (this.dist.get(i).get(i) < 0) {
                return true;
            }
        }
        return false;
    }
}

module.exports = FloydWarshall;

File: frontend/src/components/benchmark/content/floyd_warshall/implementations/floyd_warshall.java
import java.util.*;

public class FloydWarshall {
    private Map<Integer, Set<Integer>> nodes = new HashMap<>();
    private Map<Integer, Map<Integer, Integer>> dist = new HashMap<>();
    private Map<Integer, Map<Integer, Integer>> nextNode = new HashMap<>();

    public void addEdge(int u, int v, int weight) {
        nodes.putIfAbsent(u, new HashSet<>());
        nodes.putIfAbsent(v, new HashSet<>());
        nodes.get(u).add(v);
        nodes.get(v).add(u);

        dist.putIfAbsent(u, new HashMap<>());
        dist.putIfAbsent(v, new HashMap<>());
        dist.get(u).put(v, weight);
        dist.get(v).put(u, weight);

        nextNode.putIfAbsent(u, new HashMap<>());
        nextNode.putIfAbsent(v, new HashMap<>());
        nextNode.get(u).put(v, v);
        nextNode.get(v).put(u, u);
    }

    public void addDirectedEdge(int u, int v, int weight) {
        nodes.putIfAbsent(u, new HashSet<>());
        nodes.putIfAbsent(v, new HashSet<>());
        nodes.get(u).add(v);

        dist.putIfAbsent(u, new HashMap<>());
        dist.putIfAbsent(v, new HashMap<>());
        dist.get(u).put(v, weight);

        nextNode.putIfAbsent(u, new HashMap<>());
        nextNode.putIfAbsent(v, new HashMap<>());
        nextNode.get(u).put(v, v);
    }

    public void buildFromEdges(List<int[]> edges, boolean directed) {
        for (int[] edge : edges) {
            if (directed) {
                addDirectedEdge(edge[0], edge[1], edge[2]);
            } else {
                addEdge(edge[0], edge[1], edge[2]);
            }
        }
    }

    private void initialize() {
        List<Integer> allNodes = getAllNodes();

        for (int u : allNodes) {
            dist.putIfAbsent(u, new HashMap<>());
            nextNode.putIfAbsent(u, new HashMap<>());

            for (int v : allNodes) {
                if (u == v) {
                    dist.get(u).put(v, 0);
                } else if (!dist.get(u).containsKey(v)) {
                    dist.get(u).put(v, Integer.MAX_VALUE);
                }
            }
        }
    }

    private List<Integer> getAllNodes() {
        return new ArrayList<>(nodes.keySet());
    }

    public void allPairsShortestPaths() {
        initialize();
        List<Integer> allNodes = getAllNodes();

        for (int k : allNodes) {
            for (int i : allNodes) {
                for (int j : allNodes) {
                    if (dist.get(i).get(k) != Integer.MAX_VALUE &&
                        dist.get(k).get(j) != Integer.MAX_VALUE &&
                        dist.get(i).get(k) + dist.get(k).get(j) < dist.get(i).get(j)) {
                        dist.get(i).put(j, dist.get(i).get(k) + dist.get(k).get(j));
                        nextNode.get(i).put(j, nextNode.get(i).get(k));
                    }
                }
            }
        }
    }

    public int shortestPath(int u, int v) {
        if (!dist.containsKey(u) || !dist.get(u).containsKey(v)) {
            return Integer.MAX_VALUE;
        }
        return dist.get(u).get(v);
    }

    public List<Integer> getPath(int u, int v) {
        if (!nextNode.containsKey(u) || !nextNode.get(u).containsKey(v)) {
            return new ArrayList<>();
        }

        List<Integer> path = new ArrayList<>();
        path.add(u);

        while (u != v) {
            u = nextNode.get(u).get(v);
            path.add(u);
        }

        return path;
    }

    public Map<Integer, Map<Integer, Integer>> getAllDistances() {
        return dist;
    }

    public boolean hasNegativeCycle() {
        List<Integer> allNodes = getAllNodes();

        for (int i : allNodes) {
            if (dist.get(i).get(i) < 0) {
                return true;
            }
        }
        return false;
    }
}